Enable GC Regions on Mac

mangod9 opened the pull request:
Regions will be enabled for macOS, but still disabled for iOS and tvOS builds for now (will investigate the test failures on those platforms and try to enable in subsequent PRs).

mangod9 committed:
try enabling regions on Mac again

jkotas commented:
It is ok to try and the CI may even pass if the run does not happen to generate crash dump along the way. We need a positive confirmation that enabling regions GC won't kill post-mortem diagnostic on mac. Have we done anything to resolve the problem with prohibitively large crash dumps on mac?

mangod9 commented:
the last time we looked we werent able to repro the large dumps locally. Believe @janvorli had tried it too, given lower priority we never investigated it completely. Hoping we get to root cause this time around.

mangod9 commented:
I have validated that with large memory reservation size with regions the dump generated on MacOS is about 100mb larger (so it's only about 2% more) -- not order of magnitude larger we thought it would be. Need to figure out failures in iOS/tvOS NativeAOT failures before we can enable.

mangod9 committed:
Merge branch 'main' into enable_regions_mac

mangod9 committed:
dont enable for iOS and tvOS

mangod9 committed:
Merge branch 'enable_regions_mac' of http://github.com/mangod9/runtime into enable_regions_mac

mangod9 committed:
update binary size due to regions

mangod9 commented:
excluding iOS and tvOS seems to get things to work. Will need to figure out if reducing the reservation size would get regions to work on those platforms too.

mangod9 committed:
update comment

mangod9 committed:
Merge branch 'main' into enable_regions_mac

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -140,11 +140,11 @@ inline void FATAL_GC_ERROR()
 // This means any empty regions can be freely used for any generation. For
 // Server GC we will balance regions between heaps.
-// For now disable regions for standalone GC and macOS builds
+// For now disable regions for standalone GC and iOS and tvOS builds
```
```suggestion
// For now disable regions for iOS and tvOS builds
```
might as well take this opportunity to fix this comment since we do have a standalone GC with regions.

Maoni0 commented on src/tests/nativeaot/SmokeTests/HardwareIntrinsics/Program.cs:
```diff
@@ -22,7 +22,7 @@ static int Main()
         long lowerBound, upperBound;
         lowerBound = 1200 * 1024; // ~1.2 MB
-        upperBound = 1600 * 1024; // ~1.6 MB
+        upperBound = 1700 * 1024; // ~1.7 MB
```
what's the reason for this change?

mangod9 replied:
this test checks for size_on_disk of the test binary. Regions adds more code which puts it over the test threshold.

jkotas commented:
> disabled for iOS and tvOS builds for now (will investigate the test failures on those platforms and try to enable in subsequent PRs).

I do not think this is a good plan. Analyzing failures and debugging iOS/tvOS is much harder than debugging macOS. If a segments-specific issue gets introduced in the meantime, it will be a lot harder and much more time consuming to diagnose it.

mangod9 commented:
ok, will investigate failures and enable for them too.

mangod9 committed:
make regions_ranage 2x physical_mem for WKS

filipnavara commented on src/coreclr/gc/gc.cpp:
```diff
@@ -49318,8 +49318,13 @@ HRESULT GCHeap::Initialize()
             // If no hard_limit is configured the reservation size is min of 1/2 GetVirtualMemoryLimit() or max of 256Gb or 2x physical limit.
             gc_heap::regions_range = max((size_t)256 * 1024 * 1024 * 1024, (size_t)(2 * gc_heap::total_physical_mem));
+#else // MUTLIPLE_HEAPS
+            // If no hard_limit is configured the reservation size is min of 1/2 GetVirtualMemoryLimit() or max of 256Gb or 2x physical limit.
```
Update the comment?

mangod9 committed:
update reservationSize on iOS

on my local testing looks like reservation of 1/4th of physical memory seems to succeed.

mangod9 committed:
fixing typo

mangod9 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -49318,8 +49318,18 @@ HRESULT GCHeap::Initialize()
+#else // MUTLIPLE_HEAPS
+#if defined(TARGET_IOS) || defined(TARGET_TVOS)
+            // On iOS and tvOS only try to reserve 1/4 of physical memory
+            gc_heap::regions_range = (size_t)(gc_heap::total_physical_mem / 4);
```
`physical_mem / 4` was required to get the iOS sample to work on my device. Looks like that works for CI as well. Haven't found any specific guidance online around specific policy here. CoPilot claims 2x to 4x reservation should be allowed on iOS but that doesn't seem to be the case. @vitek-karas, any guidance here?

jkotas replied:
How much is `physical_mem / 4` in absolute numbers?

jkotas replied:
https://alwaysprocessing.blog/2022/02/20/size-matters is a great description of iOS virtual address space limitations. The available address space is limited and precious, very similar to the situation on 32-bit OSes.

mangod9 replied:
> How much is `physical_mem / 4` in absolute numbers?

on my iPhone 16 Pro its `2gb` (8gb total physical memory). Agree that we should keep iOS on segments for now.

jkotas replied:
Then the question is whether we want to keep macOS on segments too to get a real-world test coverage for segments on 64-bit architectures.

mangod9 replied:
Don't have a strong preference since Mac has always been on segments. However, given the overall memory policy differences do we feel that iOS specific issues would be repro-able on Mac to help with diagnosing?

mangod9 commented:
closing this PR for now, given the challenges with iOS probably reasonable to leave Apple platforms on segments for now. We will evaluate again if there is a strong case of regions helping on MacOS.
